src/backend/distributed/planner/recursive_planning.c

```diff
@@ -359,6 +364,18 @@ RecursivelyPlanSubqueriesAndCTEs(Query *query, RecursivePlanningContext *context
 		RecursivelyPlanLocalTableJoins(query, context);
 	}
 
+	/*
+	 * Similarly, logical planner cannot handle outer joins when the outer rel
+	 * is recurring, such as "<recurring> LEFT JOIN <distributed>". In that case,
+	 * we convert distributed table into a subquery and recursively plan inner
+	 * side of the outer join. That way, inner rel gets converted into an intermediate
+	 * result and logical planner can handle the new query since it's of the from
+	 * "<recurring> LEFT JOIN <recurring>".
+	 *
+	 * See DeferredErrorIfUnsupportedRecurringTuplesJoin for the supported join
+	 * types.
+	 */
```

Comment:
note that calling RecursivelyPlanRecurringTupleOuterJoins is not cheap. Should we skip if there are no outer joins at all?

We could either do like `joinRestriction->hasSemiJoin` flag, or just iterate over context->joinRestrictContext here.